apps/meteor/server/services/user/service.ts

```diff
@@ -0,0 +1,333 @@
+		}
+
+		delete options.globalRoles;
+
+		if (doc.services && !doc.services.password) {
+			const defaultAuthServiceRoles = parseCSV(settings.get('Accounts_Registration_AuthenticationServices_Default_Roles') || '');
+
+			if (defaultAuthServiceRoles.length > 0) {
+				globalRoles.push(...defaultAuthServiceRoles);
+			}
+		}
+
+		const roles = getNewUserRoles(globalRoles);
+		const { name = this.getNameFromProfile(options.profile) } = doc;
+
+		const user: InsertionModel<IUser> = {
+			createdAt: new Date(),
+			_id: Random.id(),
+			status: UserStatus.OFFLINE,
+
+			...doc,
+			type: doc.type || 'user',
+			services: {
+				...(doc.services || {}),
+				...(settings.get<boolean>('Accounts_TwoFactorAuthentication_By_Email_Auto_Opt_In') ? { enabled: true, changedAt: new Date() } : {}),
+			},
+			active: doc.active ?? !settings.get('Accounts_ManuallyApproveNewUsers'),
+			...(!!name && { name }),
+			roles: [],
+		};
+
+		if (user.services) {
+			const verified = settings.get<boolean>('Accounts_Verify_Email_For_External_Accounts');
+
+			for (const service of Object.values(user.services) as Record<string, any>[]) {
+				if (!user.name && typeof service === 'object' && ('name' in service || 'username' in service)) {
+					user.name = service.name || service.username;
+				}
+
+				if (!user.emails && service.email) {
+					user.emails = [
+						{
+							address: service.email,
+							verified,
+						},
+					];
+				}
+
+				Object.keys(service).forEach((key) => {
+					const value = service[key];
+					// If there's any value that was encrypted without an userId, decrypt it and re-encrypt with the id.
+					if (OAuthEncryption?.isSealed(value)) {
+						service[key] = OAuthEncryption.seal(OAuthEncryption.open(value), user._id);
+					}
+				});
+			}
+		}
+
+		await beforeCreateUserCallback.run(options, user);
+
+		if (!user.active && settings.get('Accounts_ManuallyApproveNewUsers')) {
+			await this.notifyAdminOfNewUser(options, user);
+		}
+
+		await callbacks.run('onCreateUser', options, user);
+
+		// App IPostUserCreated event hook
+		await Apps.triggerEvent(AppEvents.IPostUserCreated, { user, performedBy: await safeGetMeteorUser() });
+
+		await this.validateNewUser(user);
+
+		const _id = await this.insertOne(user);
+
+		const hasAdmin = await Users.findOneByRolesAndType('admin', 'user', { projection: { _id: 1 } });
+		if (!roles.includes('admin') && !hasAdmin) {
+			roles.push('admin');
+		}
+
+		await addUserRolesAsync(_id, roles);
+
+		const newUser = await Users.findOne({
+			_id,
+		});
+
+		if (!newUser) {
+			throw new Error('error-user-not-found');
+		}
+
+		if (!hasAdmin && roles.includes('admin') && settings.get('Show_Setup_Wizard') === 'pending') {
+			await Settings.updateValueById('Show_Setup_Wizard', 'in_progress');
+		}
+
+		if (newUser.username) {
+			await this.addUserToDefaultChannels(newUser, options);
+			await this.callAfterCreateUser(newUser);
+			await this.setDefaultAvatar(newUser);
```

Comment:
Can this use a Promise.all?

Reply:
I didn't want to affect the order in which they are run - yet.